Develop pk -  users form/create/edit/delete

piotrkow95 opened the pull request:
-Added users feature, adding/editing/deleting, to be consistent with branch functionality
-Home button now directs to main page

@MasterMario4 @MNowicki87 Jak wg. Was należałoby podzielić funkcjonalności na controllery funkcjonalności usera adminowych i zwykłego usera? Obecnie wrzuciłem, żeby działał sam mechanizm dodawania użytkowników i wyświetlania, lecz jest to dostęp wyłącznie dla admina, a user powinien być w stanie wyświetlić swój profil za pomocą przycisku "Profile", jak sądzę?

piotrkow95 committed:
//Done similarly to branches to be consistent
-Added user register functionality, edit user, delete user (TODO refactor to make it not only admin usage)
-Home button defined in components.html now redirects to home page

piotrkow95 committed:
//Done similarly to branches to be consistent
-Added user register functionality, edit user, delete user (TODO refactor to make it not only admin usage)
-Home button defined in components.html now redirects to home page

Mbojanow commented on src/main/java/pl/sdacademy/carrental/controllers/ClientController.java:
```diff
@@ -0,0 +1,63 @@
+    @PostMapping("/edit/{id}")
+    public String handleClientEdit(@PathVariable final Long id, @ModelAttribute(name = "branchForm") final ClientForm clientForm) {
+        clientService.updateClient(id, clientForm);
+        return "redirect:/admin/clients/";
```
/ na koncu niepotrzebny?

Mbojanow commented on src/main/java/pl/sdacademy/carrental/repositories/ClientRepository.java:
```diff
@@ -0,0 +1,8 @@
+import pl.sdacademy.carrental.domain.Client;
+
+
+public interface ClientRepository extends JpaRepository<Client,Long> {
```
Client, spacja Long

piotrkow95 replied:
Intelij mi tego nie poprawia automatycznie, zarówno ze spacją jak i bez po ctrl+alt+L zostaje, muszę pogrzebać w opcjach

Mbojanow commented on src/main/java/pl/sdacademy/carrental/services/ClientService.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public void delete(final Long id) {
+        final Client client = clientRepository.findById(id).orElseThrow();
```
orElseThrow -> lepiej konkretny wyjątek?

Mbojanow commented on src/main/java/pl/sdacademy/carrental/services/ClientService.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private Address createAddressFromClientForm(final ClientForm clientForm){
+        final Address address = Address.builder()
```
opcjonalnie -> wyrzucić te mapowanie do osobnego beanu

Mbojanow commented on src/main/resources/templates/client-edit.html:
```diff
@@ -0,0 +1,60 @@
+        <p>Edit client with following data:</p>
+        <label>
+            First Name
+            <input type="text" th:field="*{firstName}">
```
nigdy nie pamiętam -> czy th:value nie jest potrzebne do początkowej wartości?

piotrkow95 replied:
W sensie, początkowa wartość, chodzi o to żeby zapisywało początkowo wpisaną wartość? Z tego co kojarzę to th:field binduje wartości wpisane. W sensie jak się edytuje to widać to co zostało wpisane już w obecnej formie.

MNowicki87 replied:
chyba nie jest niezbędne (`th:field` powinno w tym wypadku wystarczyć)

MNowicki87 replied:
tutaj też brakuje atrybutu `required` w inputach

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/services/ClientService.java:
```diff
@@ -0,0 +1,88 @@
+                .build();
+        return addressRepository.save(address);
+    }
+    // I don't know if saving both branch and client address to the same repository is a good idea
+    // Split, refactor AddressRepository to ClientAddressRepository and BranchAddressRepository?
+    // same with Address entity? What do you think?
```
Both respresent same set of data - address, consisting of identical pieces / fields. Having two tables, entity classes and repos serve one and the same purpose contradicts DRY principle.
What if, at some point, we will have to store adresses of employees, or if cars could be delivered to clients (to addresses other than the ones used for billing)? In such cases we would have to store and handle these pieces of data separately as well. In my opinion it is best to keep the addresses table, address entity class and Address Repo as it is - after all a *branch address* is no different from *client billing address* or _prospective_ delivery address in term of its structure.

piotrkow95 replied:
Ouch, didn't delete comment about Adress. I commented that but resolved already during writing.

The question about UserController and AdminSideUserController stays though.

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/controllers/ClientController.java:
```diff
@@ -0,0 +1,63 @@
+import pl.sdacademy.carrental.model.ClientForm;
+import pl.sdacademy.carrental.services.ClientService;
+
+// (PIOTR) i'm not sure how to split admin-only methods and those available for users. From what i understand with Profile button you can
+// check your own profile as a certain user, but users list should be available only for admins? So different controller
+// for everything single user related?
```
I would advocate for having a separate controller for administrating clients from an admin perspective and another for client to be able to manage his own account/profile. To me it would be cleaner and easier to maintain. Admin limited functionality should be accessible only through urls starting with `…/admin/**`, clients _dashboard_ or _panel_, should have a completely separate base url.

MNowicki87 replied:
```
// (PIOTR)[...] From what i understand with Profile button you can
// check your own profile as a certain user [...]
```
BTW don't let the current links in the Navbar mislead you - that's a default boostrap template (nothing there is intentional 😄 ).

piotrkow95 replied:
hah, default templates for the win :P I'll create class for user functionalities but i will not move methods between them yet, left for discussion. IMO from current functionalities only showing list of all clients and deleting them should be admin-only

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/controllers/ClientController.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+
+    @GetMapping("/")
```
("/") do usunięcia

MNowicki87 commented on src/main/resources/templates/client-edit.html:
```diff
@@ -0,0 +1,60 @@
+<!DOCTYPE html>
+<html lang="en" xmlns:th="http://www.thymeleaf.org">
+<head>
+    <meta charset="UTF-8">
```
`<div th:replace="_components :: headTags"></div>`

MNowicki87 commented on src/main/resources/templates/client-create.html:
```diff
@@ -0,0 +1,69 @@
+<!DOCTYPE html>
+<html lang="en" xmlns:th="http://www.thymeleaf.org">
+<head>
+    <meta charset="UTF-8">
```
`<div th:replace="_components :: headTags"></div>`

MNowicki87 commented on src/main/resources/templates/client-edit.html:
```diff
@@ -0,0 +1,60 @@
+
+        <label>
+            Email
+            <input type="text" th:field="*{email}">
```
atrybut `type="email"` dodaje walidację na poziomie widoku i nie pozwoli wysłać formatki wyrzucającej stacktrace 😉

MNowicki87 commented on src/main/resources/templates/client-create.html:
```diff
@@ -0,0 +1,69 @@
+        <p></p>
+        <label>
+            Email
+            <input type="text" th:field="*{email}">
```
tutaj też `type="email"`

MNowicki87 commented on src/main/resources/templates/client-create.html:
```diff
@@ -0,0 +1,69 @@
+        </label>
+        <label>
+            Phone number
+            <input type="text" th:field="*{phoneNumber}">
```
`pattern="\d{9}"` spowoduje, że do formularza trzeba wpisać ciąg 9 cyfr, a tag `title="..."`, w przeciwnym wypadku, wyświetli odpowiednią wiadomość

MNowicki87 commented on src/main/resources/templates/client-create.html:
```diff
@@ -0,0 +1,69 @@
+    </form>
+</div>
+<p></p>
+<form method="get" th:action="@{/admin/clients}">
+    <button type="submit">Back to the client list</button>
+</form>
```
Nie wykonujesz tutaj żadnej akcji - zamiast osobnego "formularza" i przycisku w nim, użyj linku (`<a th:href:"@{/admin/clients}">Back to the client list</a>`).

--- _side note:_
Przycisków używa się do wykonania interakcji z widokiem, do przekierowań, czy odniesień nawet w obrębie jednej podstrony (np. scrollujące do elementu o konkretnym id) należy wykorzystywać odnośniki (można je ostylować, żeby wyglądały jak przyciski, jednak w HTML powinien być użyty tag `<a>`). Takie podejście poprawia tzw. _accessibility_, jest zgodne ze specyfikacją HTML5 (semantyczne użycie tagów), a przy okazji zostanie jedna, zamiast 3 linijek kodu 😉

piotrkow95 replied:
Będę o tym pamiętać, dzięki :)

MNowicki87 commented on src/main/resources/templates/client-create.html:
```diff
@@ -0,0 +1,69 @@
+    <nav th:replace="_components :: navigationBar"></nav>
+</header>
+<div class="client-create-from-container">
+    <form method="post" class="client-form" th:action="@{/admin/clients/create}" th:object="${clientForm}">
```
do każdego wymaganego inputu w formularzu (czyli w tym formularzu chyba do każdego) warto dodać atrybut `required` (chyba nie ma możliwości ustawienia tego jednorazowo dla całej formatki - ale jeśli znajdziesz sposób, to chętnie poznam)

MNowicki87 replied:
j.w.

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/services/ClientService.java:
```diff
@@ -0,0 +1,88 @@
+                .street(address.getStreet())
+                .apartment(address.getApartment())
+                .zipCode(address.getZipCode())
+                .phoneNumber(client.getPhoneNumber())
```
brakuje `.building(address.getBuilding())` i `.email(client.getEmail())`

MNowicki87 replied:
j.w.

piotrkow95 replied:
Dołożyłem IOException w miejscu gdzie trzeba należało dołożyć konkretny, ale nie jestem 100% pewny, czy jest to najodpowiedniejszy exception jaki można było wstawić.

MNowicki87 replied:
ten IOE tutaj średnio, lepiej dać dedykowany, a jeśli nie chcesz teraz tego robić, to zmienić metodę z `.findById(id)` na `.getOne(id)` - ona rzuca `EntityNotFoundException`

MNowicki87 reviewed (changes requested):
jest kilka rzeczy do poprawki ;-)

piotrkow95 committed:
//Done similarly to branches to be consistent

-Fixes according to reviews. Just force merge if everything's fine and i won't be online. To save previous time :D

MNowicki87 commented on src/main/java/pl/sdacademy/carrental/controllers/AdminSideClientController.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+
+    @GetMapping()
```
niepotrzebne nawiasy

MNowicki87 reviewed (changes requested):
w sumie już miałem robić merge, ale brakujące pola w getById(), w ClientService mnie powstrzymały…

piotrkow95 committed:
//Done similarly to branches to be consistent

-Fixes according to reviews PART 2. Just force merge if everything's fine this time hopefully :D

MNowicki87 reviewed (approved):
zatiwerdzam i merguje - ten wyjątek będzie na później 😉
